### PR TITLE
Update the redshift sql operator to accept multiple sql statements

### PR DESCRIPTION
**Pull request by chrishronek**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by uranusjr** [comment]
Pleas fix the static checks. Also, you need to add `sqlparse` to the requirements; search for `sqlparse` in the code base to find where you need to add it (it’s currently used by the Apache Drill provider.

**Comment by pohek321** [comment]
@uranusjr can you elaborate on what you mean by static checks? I'm a newbie and not quite sure what that means. I've also added `sqlparse` to the `setup.py` file (which is where it was for Apache Drill)

File: airflow/providers/amazon/aws/operators/redshift_sql.py [hunk 1]
```diff
@@ -74,5 +75,12 @@ def get_hook(self) -> RedshiftSQLHook:
...
+        sql_stmts = sqlparse.split(self.sql)
         hook = self.get_hook()
-        hook.run(self.sql, autocommit=self.autocommit, parameters=self.parameters)
+
+        with hook.get_conn() as con:
+            con.autocommit = self.autocommit
+            with con.cursor() as cursor:
+                for stmt in sql_stmts:
+                    cursor.execute(stmt)
+                    if self.autocommit is False: con.commit()
```
Comment (eladkal) [inline airflow/providers/amazon/aws/operators/redshift_sql.py]: I don't think this is the right approach.
The functionality should exist in the hook and the operator should utilize it.
The operator should not use cursor directly as you do it here.

See https://github.com/apache/airflow/pull/15533 adding similar functionality for Snowflake

Comment (chrishronek) [reply airflow/providers/amazon/aws/operators/redshift_sql.py]: @eladkal I went ahead and moved lines 81-86 to the RedshiftSQLHook as a sub method and called it in the RedshiftSQLOperator. Is this the right approach that you were envisioning?

Comment (eladkal) [reply airflow/providers/amazon/aws/operators/redshift_sql.py]: Not quite...
The function `run` comes from:
https://github.com/apache/airflow/blob/900bad1c67654252196bb095a2a150a23ae5fc9a/airflow/hooks/dbapi.py#L163
your suggestion to not use it and set `run_sql` means that now the hook has two ways to execute sql. This is not ideal.

You should make it work with `run`. You can override the `run` by having a new implementation in `RedshiftSQLHook` (Just like we do it for Snowflake in the PR I shared)


To be specific:
The operator calling `hook.run()` should stay exactly as it is. Let me also point that by removing `parameters=self.parameters` in the operator you are forcing breaking change which is something that we avoid unless absolutely necessary (and I don't think this is a case where it's necessary correct me if i'm wrong)
 So the hook needs to handle single query or multiple in the `run` function without changing how the `run` function is called from the operator.

**Comment by SamWheating** [comment]
I think that the base `DbAPIHook` already supports a list of statements:

https://github.com/apache/airflow/blob/900bad1c67654252196bb095a2a150a23ae5fc9a/airflow/hooks/dbapi.py#L163-L196

Can't we just supply a list of SQL statements to the existing operator?

**Comment by eladkal** [comment]
> Can't we just supply a list of SQL statements to the existing operator?

Yes but the goal is to support users who pass string composed of multiple query like:

`SELECT 1; SELECT 2;`
Most DBs knows how to execute it without any problem (MySQL, PostgresSQL etc..) others like Trino, Presto, Snowflake, Redshift can handle only 1 statement per query.

we can always ask users to pass it as `['SELECT 1', 'SELECT 2']` but that is not that nice and also note that you can't use it with templating. So if you want to use `my_queries.sql ` you can't do that with passing list.

**Comment by pohek321** [comment]
@eladkal I should've noticed your reference to the Snowflake one, that makes sense. I went ahead and committed something similar to the RedshiftSQLHook. Can you take a look and provide feedback?

**Comment by potiuk** [comment]
Static checks need fixing.

**Comment by eladkal** [comment]
@pohek321 are you still working on this issue?

**Comment by eladkal** [comment]
I think we should revive https://github.com/apache/airflow/pull/17329
It should solve the missing functionality from the root. Then it would be very easy to add it for RedShiftSQLHook, TrinoHook, SnowflakeHook etc..

**Comment by pohek321** [comment]
> @pohek321 can you fix the static checks? I think this change require also some unit tests

@eladkal I think I fixed the static checks. Can you confirm?

**Comment by potiuk** [comment]
> @eladkal I think I fixed the static checks. Can you confirm?

No. they still fail if you take a look.

File: airflow/providers/amazon/aws/hooks/redshift_sql.py [hunk 2]
```diff
@@ -133,3 +135,55 @@ def get_conn(self) -> RedshiftConnection:
...
+    def get_autocommit(self, conn):
+        return getattr(conn, 'autocommit_mode', False)
+
+    def run(
+        self,
+        sql,
+        autocommit: bool = False,
+        parameters: Optional[Union[Sequence[Any], Dict[Any, Any]]] = None,
+        handler: Optional[Callable] = None,
+    ) -> None:
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/hooks/redshift_sql.py:151]: This return type annotation is wrong

**Comment by vincbeck** [comment]
For information, there is also the operator `RedshiftDataOperator` ro tun SQL query against redshift. It uses `redshift-data` service. As of today, it only support single query but it can be easily extended to multiple queries using [batch-execute-statement](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/redshift-data.html#RedshiftDataAPIService.Client.batch_execute_statement). I am not saying it should be done as part of this PR (actually I think the opposite) but I just wanted to share the knowledge.

- If the goal here is to add multiple query support to RedshiftSQLOperator, then ignore this comment
- If the goal here is to add a feature to run multiple SQL query against Redshift. It might be easier doing it using `redshift-data`

**Comment by hyungryuk** [comment]
Do we have any progress here?

**Comment by potiuk** [comment]
Did you see the merged PR? What does it telly ou @hyungryuk ? Can you asses the state from it?

**Comment by eladkal** [comment]
> Do we have any progress here?
If you will check the code in main branch you will see that this PR is not needed any more and this operator is deprecated.

**Comment by hyungryuk** [comment]
Oh it is deprecated. it's my mistake thanx @eladkal 
So. now we can split multiple queries using **split_statements** parameter. I get it now

**Pull request closed without merging**